### ISSUE TITLE
Version 3.11: Urgent! Please include new 'in person/online' setting in import functionality and template

### ISSUE THREAD

**Issue by OSAwebsite** [code4recovery/12-step-meeting-list#408]
In the new 3.;10 version, the checkbox for Location Temporarily Closed has disappeared. 
Instead, a new radio button has been added  in the Location Information section, to indicate if a meeting is in person, online or both.  The code then decides, based on that radio button field how to tag the meeting.

After the upgrade, the tags applied to existing listings are **randomnly incorrect**. This is **not** acceptable.
It means that as an admin I have to go through every single one of my hundreds of listings in the Wordpress interface to check and correct it. It would not be such an issue if the new radio button option was now included in the import template. But that template looks as if it has not been updated for a long time, e.g. the Location temporarily closed option is still present. 

**Please fix the import so that admins who have large Areas to administer online can decide to do a full refresh of their meetings using the import option.**

I know that there are differences of opinion about all of our preferred methods (import full set regularly from a non-website datasource, strictly use the Wordpress UX to edit meeting listings and other). But we all do our service in the way that works best for us. So it would be wonderful to **keep the import updated** with whatever changes happen in the plugin.

**I cannot justify to upgrade to the latest version as long as this is not fixed.**

I find this new logic a bit wonky with too much chance of error for the simple reason that the conclusions in the code are implicit rather than explicit. The less verbal explanations are needed for new features, the more intuitive they are.

So I am just wondering why the decision was made to completely change the logic and make it worse instead of better.
Is the show/hide map decision related to geocoding costing too much money?

Thinking of solutions, I would much prefer to have one or more 'settings' sections in the back-end, that the admins can use to impact what is shown or hidden on their meeting pages. The code can use the admin choices to come to transparent choices, and make those choices compatible with the AA meeting guide. The new radio button choice might be a first step to get there, but we need more.

**A section 'meeting information'** would have check boxes (=multiple options) to indicate language, special target groups (men, women, gay, transgender etc). We already have that in place and we are all familiar with the process.

**A section 'meeting access'** could simply show check boxes (=multiple options) for online, in person and hybrid.

**A section 'display location'** could simply show check boxes for display/hide address, strike through address, show/hide map.

Here is how I interpret what I have been told about the code rules in this version and where I see the many places that can lead to errors and confusion on the actual webpages:

**ONLINE ONLY MEETING**
Listing needs:
1.  a 'No' in the new radio button field 
2. A valid URL for people to access the online meeting
3. Something in the address section to ensure the website can show a general region; address does not have to specific

Results on this front end content:
1. Meeting tag: In-person
2. Address of the meeting: shows it as strike through
3. The map: hidden
4. Join with... button: displayed, provided there is valid info in the URL field; if there is none

Chance of confusion and errors:
1. If I enter an invalid URL or if there is no URL to show because people have to email or phone to get access details, the meeting now reverts to 'in-person' and there is nothing I can do about that as admin. It is a realistic scenario and it causes an error that I have no influence on as the admin
2. I can enter a bogus address that may or may not be recognised as a geographical location, however, the field region can contain something else entirely and is of no consequence for the plugin (apparently). No error code when the address is not found. Listing tag is INCORRECTLY changed to Temporarily Inactive
3. Visitors might be confused when the address could not be verified and the details page shows up very empty. They may think: the map is not working! Something is wrong!

**IN PERSON ONLY MEETING**
Listing needs:
1.  a 'Yes' in the new radio button field 
2. Nothing in the URL field
3. Valid and geocode verifiable meeting address in the address section

Acts on this front end content:
1. Meeting tag: in-person only
2. The address of the meeting: is displayed, provided it is geocode verifiable
3. The map: displayed
4. Join with... button: hidden 

Chance of confusion and errors:
1. When the address cannot be verified, the plugin throws an error and the meeting will not update: fair enough
2. When the admin also still has something in the URL field that is valid, the tag of the meeting suddenly changes to 'Online' and 'in-person' and it will display the 'Join with...' button, while this is incorrect and confusing

**HYBRID MEETING** 
Listing needs:
1.  a 'Yes' in the new radio button field 
2. A valid URL for people to access the online meeting
3. Valid and verifiable meeting address in the address section

Results on this front end content:
1. Meeting tag: 'in-person' and 'online', provided there is a valid URL; if not, it just shows 'in person', which is an error
2. The address: displayed
3. The map: displayed provided it is geocode verifiable
4. Join with... button: displayed provided the URL is valie

Chance of confusion and errors:
1. There are different models of hybrid meetings. Physical meetings can be broadcast and attended, having the same people online as there are physically. Or there may be two meetings: one physical, one online. So when we get the tags in-person and online, the meeting information box needs to contain an exact explanation. The admins can do that in the box underneath the checkbox. However, if they go through all that trouble and an error in address or URL happens (progress not perfection, remember, we all do make mistakes like that) the tags will change and cause confusion for visitors

In summary: I would at the very least request that the new radio button choice is integrated in the meeting template, which will be a good opportunity to double-check that it in fact has everything that we need. This gives us choices and that is in accordance with the content of our programs, is it not?

In future upgrades I would love to see more explicit admin choices that can be set with predictable and realistic outcomes on the websites. It may be something that takes a bit of extra time to set up in the code, but it will be worth it because it will be very simple to expand on and to debug for the development team. Right now, the consequences of certain meeting information content is a guessing game. 

Thanks for considering all this.
Nicky

**Comment by tech2serve** [comment]
We are discussing the importer and whether or not changes are warranted. Below I’ll share why I think it is working essentially the same as it was, that the GUI we’ve changed doesn’t fundamentally change the logic, and that some of your concerns may be a result of improvements to data validation and specification compliance.

> It means that as an admin I have to go through every single one of my hundreds of listings in the Wordpress interface to check and correct it. It would not be such an issue if the new radio button option was now included in the import template. But that template looks as if it has not been updated for a long time, e.g. the Location temporarily closed option is still present.

I’m glad you’re testing this extensively on your dev site first. As best I can tell, the lack of strictness regarding data in previous versions may be the real cause of the issue. Given the complexity inherent to tracking and supporting multiple attendance options vice just one means increased compliance and validation was necessary. Because of this, it may be necessary to massage the data in mass before migrating.

The TC type remains in v3.11 because it is necessary for compatibility with the Meeting Guide specification. Users can take advantage that with the importer. 

In short, the issue with simply adding the field to the importer as you’ve proposed is that we then have to add code to validate other fields. For example, if the admin indicates a meeting should have the attendance option of _online_ but then doesn’t include conference URL or phone, that meeting is not valid and shouldn’t publish, or should throw an error. A simpler and more accurate approach is that if the admin wants the meeting to include the _online_ attendance option, the admin includes a conference URL/phone in that row. 

If the admin wants the meeting to reflect an attendance option of _in person_, they should include a complete specific address. 

If we don’t handle it this way, we’d have to write and maintain more code to validate the address on import. This simply isn’t necessary as the information necessary for online or in-person meetings must be entered anyway.

Handling meetings that are temporarily inactive is not quite as straightforward, but it isn’t any different than it is for previous versions released since the pandemic. For that, the admin should continue to include TC as a type. 

> 
> **Please fix the import so that admins who have large Areas to administer online can decide to do a full refresh of their meetings using the import option.**

At this time, I don’t see the importer as broken but we’ll continue to investigate.

There are roughly three logic decisions to be made in the spreadsheet, and this really isn’t drastically different that previous versions:

1. If the meeting is to be available to users _online_, include the conference URL or phone. If the meeting is to be *only* available online, use an approximate address.
2. If the meeting is to be available to users _in person_, include a specific non-approximate address.
3. If the meeting location is temporarily not available, add the TC type along with the specific non-approximate address, and the meeting should be flagged as inactive.

> 
> I know that there are differences of opinion about all of our preferred methods (import full set regularly from a non-website datasource, strictly use the Wordpress UX to edit meeting listings and other). But we all do our service in the way that works best for us. So it would be wonderful to **keep the import updated** with whatever changes happen in the plugin.
>

I remain confident we have the code correct based on the Meeting Guide specification and our intentions. If we see the data which makes it clear we have a bug, we’ll work diligently to fix it. I sure won’t ever claim we’re bug free.
 
> **I cannot justify to upgrade to the latest version as long as this is not fixed.**

This is your call. If you feel this strongly, you may want to fork the repo based on the v3.10 release and maintain it or morph it as you chose.
> 
> Here is how I interpret what I have been told about the code rules in this version and where I see the many places that can lead to errors and confusion on the actual webpages:
> 
> **ONLINE ONLY MEETING**
> Listing needs:
> 
> 1. a 'No' in the new radio button field
> 2. A valid URL for people to access the online meeting
> 3. Something in the address section to ensure the website can show a general region; address does not have to specific

Address *should not* be specific. A specific address creates the potential of confusing a user because they believe they can go attend in person. The address *should be* approximate in order to support region affiliation and time zone capturing.

> 
> Results on this front end content:
> 
> 1. Meeting tag: In-person
> 2. Address of the meeting: shows it as strike through
> 3. The map: hidden
> 4. Join with... button: displayed, provided there is valid info in the URL field; if there is none

I suspect this is happening because you have a specific address but the plugin has been told the meeting isn’t meeting in person, so it labels it as inactive. 

> 
> Chance of confusion and errors:
> 
> 1. If I enter an invalid URL or if there is no URL to show because people have to email or phone to get access details, the meeting now reverts to 'in-person' and there is nothing I can do about that as admin. It is a realistic scenario and it causes an error that I have no influence on as the admin

This is probably happening because you have a non-approximate address. Change it to city and state, and the problem should clear up.

> 2. I can enter a bogus address that may or may not be recognised as a geographical location, however, the field region can contain something else entirely and is of no consequence for the plugin (apparently). No error code when the address is not found. Listing tag is INCORRECTLY changed to Temporarily Inactive
> 3. Visitors might be confused when the address could not be verified and the details page shows up very empty. They may think: the map is not working! Something is wrong!

Entering city and state (as the approximate address) will geocode and be valid. The plugin won’t display a map because there is no place for a user to map to. The TSMLui frontend currently contains an option to display an image in lieu.

> 
> **IN PERSON ONLY MEETING**
> Listing needs:
> 
> 1. a 'Yes' in the new radio button field
> 2. Nothing in the URL field
> 3. Valid and geocode verifiable meeting address in the address section
> 
> Acts on this front end content:
> 
> 1. Meeting tag: in-person only
> 2. The address of the meeting: is displayed, provided it is geocode verifiable
> 3. The map: displayed
> 4. Join with... button: hidden
> 
> Chance of confusion and errors:
> 
> 1. When the address cannot be verified, the plugin throws an error and the meeting will not update: fair enough
> 2. When the admin also still has something in the URL field that is valid, the tag of the meeting suddenly changes to 'Online' and 'in-person' and it will display the 'Join with...' button, while this is incorrect and confusing
> 

If you are entering something in the conference URL that is valid, that tells the plugin the meeting is hybrid. If you want an in-person only meeting, leave the conference URL empty. Don’t use it for other purposes. 
 
> **HYBRID MEETING**
> Listing needs:
> 
> 1. a 'Yes' in the new radio button field
> 2. A valid URL for people to access the online meeting
> 3. Valid and verifiable meeting address in the address section
> 
> Results on this front end content:
> 
> 1. Meeting tag: 'in-person' and 'online', provided there is a valid URL; if not, it just shows 'in person', which is an error

If you don’t include a valid URL then it isn’t a hybrid meeting. 

> 2. The address: displayed
> 3. The map: displayed provided it is geocode verifiable
> 4. Join with... button: displayed provided the URL is valie
> 
> Chance of confusion and errors:
> 
> 1. There are different models of hybrid meetings. Physical meetings can be broadcast and attended, having the same people online as there are physically. Or there may be two meetings: one physical, one online. So when we get the tags in-person and online, the meeting information box needs to contain an exact explanation. The admins can do that in the box underneath the checkbox. However, if they go through all that trouble and an error in address or URL happens (progress not perfection, remember, we all do make mistakes like that) the tags will change and cause confusion for visitors
> 

We only support one form of hybrid meeting: A meeting for which a single meeting is available both in person and online at the same time and with the same chair person.

If there are two meetings (one online and one in person) then those are two different meetings, and they should be in your database as two separate meetings.

> In summary: I would at the very least request that the new radio button choice is integrated in the meeting template, which will be a good opportunity to double-check that it in fact has everything that we need. This gives us choices and that is in accordance with the content of our programs, is it not?
> 
> In future upgrades I would love to see more explicit admin choices that can be set with predictable and realistic outcomes on the websites. It may be something that takes a bit of extra time to set up in the code, but it will be worth it because it will be very simple to expand on and to debug for the development team. Right now, the consequences of certain meeting information content is a guessing game.
> 
> Thanks for considering all this.
> Nicky

**Comment by kiyote33** [comment]
Lots to like about 3.11 with the new UI.  I too would like to see more local admin control of some features on the import & settings page as not everyone will be 100% ready for all the new features as implemented.  Allowing each website to experiment and integrate at their own speed is a win/win approach.

**Comment by OSAwebsite** [comment]
kiyote33: good point. And better expressed than I did. I get so wordy in my quest to be transparant. But you hit the nail on the head: admin control. And also: when doing an upgrade without adjusting the data, knowing what to change in order for things to be displayed in the front end correctly.
I am currently working on a little table that shows what is needed in the old (current) version in terms of settings, in order for the upgrade to show what we need it to show. It's easy to get things confused and I need that reminder for myself.
When I have that done, I can post it here for people who would like to see it.

**Comment by tech2serve** [comment]
For the reasons cited by me above, this request will not be supported. In summary, the requested feature creates ambiguity in the data which is a bad thing.
